### DISCUSSION TITLE
can model-viewer used in android?  i have CORS-restricted problem.

### DISCUSSION THREAD

**Discussion by whitebird101** [google/model-viewer#2197] [Q&A]
can model-viewer.js used in  android app? 
when i try use model-viewer.js in my android app, and Access the JS file from local file system.
the following error occurs:
`I/chromium: [INFO:CONSOLE(0)] "Access to Script at 'file:///android_asset/model-viewer.js' from origin 'null' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource. Origin 'null' is therefore not allowed access.", source: file:///android_asset/test.html (0)`
what can i do if i want use this js in android app?

**Comment by cdata** [comment]
@whitebird101 yes, we absolutely can be used on Android. However, you must work within the sandboxing rules of the browser, and this includes cross-origin restrictions.

My recommendation is to run a local web server and host `test.html` and `model-viewer.js` there, so that you can be sure that the origins are configured as you intend, and that any CORS headers you set are correct.

For CORS issues related to serving models, I would also recommend referring to the discussion here: https://github.com/google/model-viewer/issues/1015

**Comment by whitebird101** [comment]
@cdata thanks for your reply！But, specific to my project, in many cases, there is no network support, no WIFI. so, my app must run locally。in this case, i want know that can model-viewer.js  be used on Android? without  any local web server or host.
can model-viewer.js  be used just like JQuery.js? put the JS file in local folder，and write a HTML file , import model-viewer.js  through <script> tag, and then display a 3D model through <model-viewer> tag?
thanks again!

**Reply by GiovanniTurri** [reply]
I still find this issue, but it's noticeable that an offline solution is not possible, as the model-viewer.js script can't work offline (for what is my experience doing some test).
As @cdata says, it's mandatory to use the sandbox of the browser (e.g. file:///storage/emulated/0/Android/data/com.google.chrome/file/download) with a .js file, but not with model-viewer.js.
@cdata do you have any suggestion for an offline solution?
Thank you

**Reply by elalish** [reply]
Our editor is based on MV and works fine offline. Simply open the file and set the `src` to an ObjectURL created from the Blob.

**Reply by GiovanniTurri** [reply]
Thank you!
From some tutorials I found that, adding a `<input>` (before `<model-viewer>` I think) in the HTML, a snippet like
```
const fileInput = document.querySelector('input');
fileInput.onchange = async () => {
    const file = fileInput.files[0];
    const objUrl =  URL.createObjectURL(file);
    document.querySelector('model-viewer').src = objUrl;
    URL.revokeObjectURL(objUrl);
}
```
should work, but a null is generated in "objUrl".

From the code that should be related to the MV editor I found [some code for USDZ models](https://github.com/google/model-viewer/blob/81326c7360cdeb15a883655dcc319f52529df8b4/packages/model-viewer/src/features/ar.ts#:~:text=const%20url%20%3D%20URL.createObjectURL(blob)%3B), but couldn't reproduce changing to something like
```
var exporter = new THREE.GLTFExporter();
const arraybuffer = exporter.parse(scene.modelContainer);
const blob = new Blob([arraybuffer], {
type: 'model/gltf+binary', });
```
Do you have any further suggestions?
Thank you again

**Reply by elalish** [reply]
I think your first code block is on the right track. You might just be revoking the ObjectURL before it's used (MV is very async).

**Reply by GiovanniTurri** [reply]
`fileInput.addEventListener("change", (event) => {` apparently made it work.
But after chosing the file I have the error
Fetch API cannot load blob:null/d10b770b-679a-45a1-9d7e-08853941c296. **URL scheme "blob" is not supported.**

Do you think it is this still due to a wrong asynchronous implementation on the revoking?
 
I would like to change the `src` immediately after the .glb file is loaded through the `<input>`
Thank you again


Update:
With an other script I can load the .glb from the `<input>` (so is visible on the browser) but when pressing the "Start AR" button, scene-viewer crashes immediately (doesn't open the camera) or I can see again **file=blob:null/ something** :
```
Impossible to load the webpage at address intent://arvr.google.com/scene-viewer/1.0?mode=ar_preferred&disable_occlusion=true&file=blob%3Anull%2F8e3155a3-07ed-4af9-b7ac-9cef37efc3cd#Intent;scheme=https;package=com.google.ar.core;action=android.intent.action.VIEW;S.browser_fallback_url=file%3A%2F%2F%2Fstorage%2Femulated%2F0%2FDownload%2Fmvofmin.html%23model-viewer-no-ar-fallback;end; perché:

net::ERR_UNKNOWN_URL_SCHEME
```

**Reply by elalish** [reply]
Yeah, the blob method can't work with Scene Viewer since it's a separate app; it can't share memory with the browser and has to download its own URL. This is why we default to WebXR mode, as it's much more flexible.

**Reply by GiovanniTurri** [reply]
Oh I see, so it may be impossible to pass the uploaded model using `<input>` to the Scene Viewer app from myFile.html , as I don't think there's an API for this in it. 

Being **totally offline**, it works for Chrome on Android 9 and 10!!!

For Android 11 unfortunately with `ar-modes="webxr"` anything happens, even copying the whole model-viewer.min.js script inside myFile.html (that I open on Chrome Mobile through Google Files app, so with URL `content://com.google.android.apps.nbu.files.provider/1/file%3A%2F%2F%2Fstorage%2Femulated%2F0%2FDownload%2FmyFile.html` ).
In Android 11 I'm not able to open myFile.html through a sandbox of Chrome (so with an URL like file:///storage/emulated/0/**Android/data/com.google.chrome/file**/download due to ERR_ACCESS_DENIED from O.S.), that maybe could have make WebXR work as is in localhost, right?

Also I confirm that, as model-viewer.dev/editor , it's possible to render the .glb in the browser offline!
Thank you

**Reply by elalish** [reply]
Not quite sure what issue you're hitting, but WebXR requires HTTPS. Also, WebXR is not available in WebView, but is available in WebLayer.
